### ISSUE TITLE
Transformer Math 101 | EleutherAI Blog

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11690]
Transformer Math 101 | EleutherAI Blog<br>
<br>
https://ift.tt/hnb8ylF<br>
<br>
Hailey Schoelkopf<br>
<br>

<div><div>
<div>
<p>A lot of basic, important information about transformer language models can be computed quite simply. Unfortunately, the equations for this are not widely known in the NLP community. The purpose of this document is to collect these equations along with related knowledge about where they come from and why they matter.</p>
<p><strong>Note:</strong> This post is primarily concerned with training costs, which are dominated by VRAM considerations. For an analogous discussion of inference costs with a focus on latency, check out <a href="https://kipp.ly/blog/transformer-inference-arithmetic/">this excellent blog post</a> by Kipply.</p>

<p>The basic equation giving the cost to train a transformer model is given by:</p>

<p>where:</p>
<ul>
<li> is the compute required to train the transformer model, in total floating point operations</li>
<li>
</li>
<li>
</li>
<li>
</li>
<li> is the aggregate throughput of your hardware setup (), in FLOPs</li>
<li> is the time spent training the model, in seconds</li>
<li> is the number of parameters in the transformer model</li>
<li> is the dataset size, in tokens</li>
</ul>
<p>These equations are proposed and experimentally validated in <a href="https://arxiv.org/abs/2001.08361">OpenAI’s scaling laws paper</a> and <a href="https://arxiv.org/abs/2203.15556">DeepMind’s scaling laws paper</a>. Please see each paper for more information.</p>
<p>It’s worth taking an aside and discussing the units of .  is a measure of total compute, but can be measured by many units such as:</p>
<ul>
<li>FLOP-seconds, which is in units of </li>
<li>GPU-hours, which is in units of </li>
<li>Scaling laws papers tend to report values in PetaFLOP-days, or  total floating point operations</li>
</ul>
<p>One useful distinction to keep in mind is the concept of . While GPU accelerator whitepapers usually advertise their theoretical FLOPs, these are never met in practice (especially in a distributed setting!). Some common reported values of  in a distributed training setting are reported below in the Computing Costs section.</p>
<p>Note that we use the throughput-time version of the cost equation as used in <a href="https://medium.com/@dzmitrybahdanau/the-flops-calculus-of-language-model-training-3b19c1f025e4">this wonderful blog post on LLM training costs</a>.</p>
<h2 id="parameter-vs-dataset-tradeoffs">Parameter vs Dataset Tradeoffs</h2>
<p>Although strictly speaking you can train a transformer for as many tokens as you like, the number of tokens trained can highly impact both the computing costs and the final model performance making striking the right balance important.</p>
<p><strong>Let’s start with the elephant in the room: “compute optimal” language models.</strong> Often  referred to as “Chinchilla scaling laws” after the model series in the paper that gave rise to current beliefs about the number of parameters, a compute optimal language model has a <strong>number of parameters</strong> and a <strong>dataset size</strong> that satisfies the approximation . This is optimal in one very specific sense: in a resource regime where using 1,000 GPUs for 1 hour and 1 GPU for 1,000 hours cost you the same amount, if your goal is to maximize performance while minimizing the cost in GPU-hours to train a model you should use the above equation.</p>
<p><strong>We do not recommend training a LLM for less than 200B tokens.</strong> Although this is “chinchilla optimal” for many models, the resulting models are typically quite poor. For almost all applications, we recommend determining what inference cost is acceptable for your usecase and training the largest model you can to stay under that inference cost for as many tokens as you can.</p>
<h2 id="engineering-takeaways-for-compute-costs">Engineering Takeaways for Compute Costs</h2>
<p>Computing costs for transformers are typically listed in GPU-hours or FLOP-seconds.</p>
<ul>
<li>GPT-NeoX achieves 150 TFLOP/s/A100 with normal attention and 180 TFLOP/s/A100 with Flash Attention. This is in line with other highly optimized libraries at scale, for example Megatron-DS reports between 137 and 163 TFLOP/s/A100.</li>
<li>As a general rule of thumb, you should always be able to achieve approximately 120 TFLOP/s/A100. If you are seeing below 115 TFLOP/s/A100 there is probably something wrong with your model or hardware configuration.</li>
<li>With high-quality interconnect such as InfiniBand, you can achieve linear or sublinear scaling across the data parallel dimension (i.e. increasing the data parallel degree should increase the overall throughput nearly linearly). Shown below is a plot from testing the GPT-NeoX library on Oak Ridge National Lab’s Summit supercomputer. Note that V100s are on the x-axis, while most of the numerical examples in the post are for A100s.</li>
</ul>
<figure><img src="https://blog.eleuther.ai/images/blog/transformer-math/neox-scaling.png#center" alt="GPT-NeoX scaling"></figure><p>Transformers are typically described in terms of their <em>size in parameters</em>. However, when determining what models can fit on a given set of computing resources you need to know <strong>how much space in bytes</strong> the model will take up. This can tell you how large a model will fit on your local GPU for inference, or how large a model you can train across your cluster with a certain amount of total accelerator memory.</p>
<h2 id="inference">Inference</h2>
<h3 id="model-weights">Model Weights</h3>
<figure><img src="https://blog.eleuther.ai/images/blog/transformer-math/dl-precisions.png#center" alt="Different floating point formats and their relative bits for precision and range"></figure><p>Most transformers are trained in <strong>mixed precision</strong>, either fp16 + fp32 or bf16 + fp32. This cuts down on the amount of memory required to train the models, and also the amount of memory required to run inference. We can cast language models from fp32 to fp16 or even int8 without suffering a substantial performance hit. These numbers refer to the size <em>in bits</em> a single parameter requires. Since there are 8 bits in a Byte, we divide this number by 8 to see how many Bytes each parameter requires</p>
<ul>
<li>In int8, </li>
<li>In fp16 and bf16, </li>
<li>In fp32, </li>
</ul>
<p>There is also a small amount of additional overhead, which is typically irrelevant to determining the largest model that will fit on your GPU. In our experience this overhead is ≤ 20%.</p>
<h3 id="total-inference-memory">Total Inference Memory</h3>
<p>In addition to the memory needed to store the model weights, there is also a small amount of additional overhead during the actual forward pass. In our experience this overhead is ≤ 20% and is typically irrelevant to determining the largest model that will fit on your GPU.</p>
<p>In total, a good heuristic answer for “will this model fit for inference” is:</p>

<p>We will not investigate the sources of this overhead in this blog post and leave it to other posts or locations for now, instead focusing on memory for model training in the rest of this post. If you’re interested in learning more about the calculations required for inference, check out <a href="https://kipp.ly/blog/transformer-inference-arithmetic/">this fantastic blog post covering inference in depth</a>. Now, on to training!</p>
<h2 id="training">Training</h2>
<p>In addition to the model parameters, training requires the storage of optimizer states and gradients in device memory. This is why asking “how much memory do I need to fit model X” immediately leads to the answer “this depends on training or inference.” Training always requires more memory than inference, often very much more!</p>
<h3 id="model-parameters">Model Parameters</h3>
<p>First off, models can be trained in pure fp32 or fp16:</p>

<p>In addition to the common model weight datatypes discussed in Inference, training introduces <strong>mixed-precision</strong> training such as <a href="https://developer.nvidia.com/automatic-mixed-precision">AMP</a>. This technique seeks to maximize the throughput of GPU tensor cores while maintaining convergence. The modern DL training landscape frequently uses mixed-precision training because: 1) fp32 training is stable, but has a high memory overhead and doesn’t exploit NVIDIA GPU tensor cores, and 2) fp16 training is stable but difficult to converge. For more information on mixed-precision training, we recommend reading <a href="https://nbviewer.org/github/tunib-ai/large-scale-lm-tutorials/blob/main/notebooks/08_zero_redundancy_optimization.ipynb">this notebook by tunib-ai</a>. Note that mixed-precision requires an fp16/bf16 and fp32 version of the model to be stored in memory, requiring:</p>
<ul>
<li>Mixed-precision (fp16/bf16 and fp32), </li>
</ul>
<p>plus an additional size  copy of the model <strong>that we count within our optimizer states</strong>.</p>
<h3 id="optimizer-states">Optimizer States</h3>
<p>Adam is magic, but it’s highly memory inefficient. In addition to requiring you to have a copy of the model parameters and the gradient parameters, you also need to keep an additional three copies of the gradient parameters. Therefore,</p>
<ul>
<li>For vanilla AdamW, 
<ul>
<li>fp32 copy of parameters: 4 bytes/param</li>
<li>Momentum: 4 bytes/param</li>
<li>Variance: 4 bytes/param</li>
</ul>
</li>
<li>For 8-bit optimizers like <a href="https://github.com/TimDettmers/bitsandbytes">bitsandbytes</a>, 
<ul>
<li>fp32 copy of parameters: 4 bytes/param</li>
<li>Momentum: 1 byte/param</li>
<li>Variance: 1 byte/param</li>
</ul>
</li>
<li>For SGD-like optimizers with momentum, 
<ul>
<li>fp32 copy of parameters: 4 bytes/param</li>
<li>Momentum: 4 bytes/param</li>
</ul>
</li>
</ul>
<h3 id="gradients">Gradients</h3>
<p>Gradients can be stored in fp32 or fp16 (Note that the gradient datatype often matches the model datatype. We see that it therefore is stored in fp16 for fp16 mixed-precision training), so their contribution to memory overhead is given by:</p>

<h3 id="activations-and-batch-size">Activations and Batch Size</h3>
<p>Modern GPUs are typically bottlenecked by memory, not FLOPs, for LLM training. Therefore activation recomputation/checkpointing is an extremely popular method of trading reduced memory costs for extra compute costs. Activation recomputation/checkpointing works by recomputing activations of certain layers instead of storing them in GPU memory. The reduction in memory depends on how selective we are when deciding which layers to clear, but Megatron’s selective recomputation scheme is depicted in the figure below:</p>
<figure><img src="https://blog.eleuther.ai/images/blog/transformer-math/activations.png#center" alt="activation memory"></figure><p>Where the dashed red line indicates the memory capacity of an A100-80GB GPU, and “present work” indicates the memory requirements after applying selective activation recomputation. See  <a href="https://arxiv.org/abs/2205.05198">Reducing Activation Recomputation in Large Transformer Models</a> for further details and the derivation of the equations below</p>
<p>The basic equation giving the memory required to store activations for a transformer model is given by:</p>

<p>where:</p>
<ul>
<li> is the sequence length, in tokens</li>
<li> is the batch size per GPU</li>
<li> is the dimension of the hidden size within each transformer layer</li>
<li> is the number of layers in the transformer model</li>
<li> is the number of attention heads in the transformer model</li>
<li> is the degree of tensor parallelism being used (1 if not)</li>
<li>We assume no sequence parallelism is being used</li>
<li>We assume that activations are stored in fp16</li>
</ul>
<p>The additional recomputation necessary also depends on the selectivity of the method, but it’s bounded above by a full additional forward pass. Hence the updated cost of the forward pass is given by:</p>

<h3 id="total-training-memory">Total Training Memory</h3>
<p>Therefore, a good heuristic answer for “will this model fit for training” is:</p>

<h2 id="distributed-training">Distributed Training</h2>
<h3 id="sharded-optimizers">Sharded Optimizers</h3>
<p>The massive memory overheads for optimizers is the primary motivation for sharded optimizers such as <a href="https://arxiv.org/abs/1910.02054">ZeRO</a> and <a href="https://engineering.fb.com/2021/07/15/open-source/fsdp/">FSDP</a>. Such sharding strategies reduce the optimizer overhead by a factor of , which is why a given model configuration may fit at large scale but OOM at small scales. If you’re looking to calculate the memory overhead required by training using a sharded optimizer, you will need to include the equations from the figure below. For some sample calculations of sharded optimization, see the following figure from the <a href="https://arxiv.org/abs/1910.02054">ZeRO</a> paper (Note that   and  are commonly denoted as ZeRO-1, ZeRO-2, ZeRO-3, respectively. ZeRO-0 commonly means “ZeRO disabled”):</p>
<figure><img src="https://blog.eleuther.ai/images/blog/transformer-math/zero_fig.png#center" alt="ZeRO illustration"></figure><figure><img src="https://blog.eleuther.ai/images/blog/transformer-math/zero_legend.png#center" alt="ZeRO legend"></figure><p>In the language of this blog post (assuming mixed-precision and the Adam optimizer):</p>

<p>Where  is just  unless pipeline and/or tensor parallelism are applied. See <a href="https://www.notion.so/Sharded-Optimizers-3D-Parallelism-9c476d020d7641a299fb6be6ae82e9f8">Sharded Optimizers + 3D Parallelism</a> for details.</p>
<p>Note that ZeRO-3 introduces a set of live parameters. This is because ZeRO-3 introduces a set of config options (<em><strong>stage3_max_live_parameters, stage3_max_reuse_distance, stage3_prefetch_bucket_size, stage3_param_persistence_threshold</strong></em>) that control how many parameters are within GPU memory at a time (larger values take more memory but require less communication). Such parameters can have a significant effect on total GPU memory.</p>
<p>Note that ZeRO can also partition activations over data parallel ranks via <strong>ZeRO-R</strong>. This would also bring the  above the tensor parallelism degree . For more details, read the associated <a href="https://arxiv.org/abs/1910.02054">ZeRO paper</a> and <a href="https://www.deepspeed.ai/docs/config-json/#activation-checkpointing">config options</a> (note in GPT-NeoX, this is the <code>partition_activations</code> flag). If you are training a huge model, you would like to trade some memory overhead for additional communication cost, and activations become a bottleneck. As an example of using ZeRO-R along with ZeRO-1:</p>

<h3 id="3d-parallelism">3D Parallelism</h3>
<p>Parallelism for LLMs comes in 3 primary forms:</p>
<p><strong>Data parallelism:</strong> Split the data among (possibly model-parallel) replicas of the model</p>
<p><strong>Pipeline or Tensor/Model parallelism:</strong> These parallelism schemes split the parameters of the model across GPUs. Such schemes require significant communication overhead, but their memory reduction is approximately:</p>

<p>Note that this equation is approximate due to the facts that (1) pipeline parallelism doesn’t reduce the memory footprint of activations, (2) pipeline parallelism requires that all GPUs store the activations for all micro-batches in-flight, which becomes significant for large models, and (3) GPUs need to temporarily store the additional communication buffers required by parallelism schemes.</p>
<h3 id="sharded-optimizers--3d-parallelism">Sharded Optimizers + 3D Parallelism</h3>
<p>When ZeRO is combined with tensor and/or pipeline parallelism, the resulting parallelism strategy forms a mesh like the following:</p>
<figure><img src="https://i.imgur.com/xMgptTN.png#center" alt="3D parallelism"></figure><p>As an important aside, the DP degree is vital for use in calculating the global batch size of training. The data-parallel degree depends on the number of complete model replicas:</p>

<p>Pipeline parallelism and tensor parallelism are compatible with all stages of ZeRO. However, it's difficult to maintain efficiency when combining pipeline parallelism with ZeRO-2/3's gradient sharding (Because ZeRO-2 shards the gradients, but pipeline parallelism accumulates them. It's possible to carefully define a pipeline schedule and overlap communication to maintain efficiency, but it's difficult to the point that DeepSpeed currently forbids it: <a href="https://github.com/microsoft/DeepSpeed/blob/v0.10.1/deepspeed/runtime/pipe/engine.py#L71)">https://github.com/microsoft/DeepSpeed/blob/v0.10.1/deepspeed/runtime/pipe/engine.py#L71)</a>. Tensor parallelism, however, is complementary to all stages of ZeRO because on each rank:</p>
<ul>
<li>ZeRO-3 gathers the full layer <strong>parameters</strong> from other ranks, processes a <strong>full</strong> input on the now-local full layer, then frees the memory that was allocated to hold the remote ranks' parameters.</li>
<li>Tensor Parallelism gathers the remote <strong>activations</strong> for the local input from other ranks, processes a <strong>partition</strong> of the input using the local layer partition, then sends the next layer's activations to remote ranks</li>
</ul>
<p>For the majority of Eleuther's work, we train with pipeline and tensor parallelism along with ZeRO-1. This is because we find ZeRO-3 to be too communication-heavy for our hardware at large scales, and instead use pipeline parallelism across nodes along with tensor parallelism within nodes.</p>
<p>Putting everything together for a typical 3D-parallel ZeRO-1 run with activation partitioning:</p>

<p>EleutherAI engineers frequently use heuristics like the above to plan efficient model training and to debug distributed runs. We hope to provide some clarity on these often-overlooked implementation details, and would love to hear your feedback at <a>contact@eleuther.ai</a> if you would like to discuss or think we’ve missed anything!</p>
<p>To cite this blog post, please use:</p>
<div><pre tabindex="0"><code>@misc{transformer-math-eleutherai,
  title = {Transformer Math 101},
  author = {Anthony, Quentin and Biderman, Stella and Schoelkopf, Hailey},
  howpublished = \url{blog.eleuther.ai/},
  year = {2023}
}
</code></pre></div>

		</div>
	</div></div><br>
<br>
<br>
<br>
<br>
<br>
via EleutherAI Blog https://ift.tt/Zna0QvU<br>
<br>
July 31, 2024 at 05:17PM
